_blogposts/compiler/string-literal-types-in-reason.mdx

```diff
@@ -0,0 +1,215 @@
+---
+author: hongbo
+date: "2020-07-28 18:00"
+previewImg:
+category: compiler
+title: Introducing string literal types in BuckleScript version 8.2
+description: |
+  Highlights of our newest changes to the internal representation 
+  and how they will benefit our users.
+---
+
+---
+
+## String literal types in BuckleScript
+
+[String literal types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#string-literal-types) was introduced by TypeScript to model
+JavaScript behavior, it's a relatively new concept since most type systems are runtime encoding agnostic. However, to smooth the user experience when writing
+bindings to existing JS API, we are introducing string literal types which is unique in several behaviors compared with TypeScript: it supports type inference,
+pattern match and can be attached to data.
+
+## Vanilla string literal types
+
+The notation in Reason for string literal types is like this \``hello`, it will be compiled into "hello", the difference is that \``hello` is given a type so that you can not mix it with other strings.
+
+Take such code snippet below as an example:
+
+```reasonml
+let encoding = (enc) =>
+  switch (enc) {
+  | `utf8 => 0
+  | `ascii => 1
+  | `utf16 => 2
+  };
+```
+
+It will be compiled into
+
+```js
+function encoding(en) {
+  if (en === "ascii") {
+    return 1;
+  } else if (en === "utf16") {
+    return 2;
+  } else {
+    return 0;
+  }
+}
+```
+
+If you pass a random encoding, e.g, `` encoding (`ucs32) ``, you get a type error:
+
+```
+This expression has type [> `ucs32 ]
+       but an expression was expected of type [< `ascii | `utf16 | `utf8 ]
+       The second variant type does not allow tag(s) `ucs32
+```
+
+Another thing you can observe from the generated JS is that since the compiler can guarantee that the input could only be `` `utf8, `ascii,`utf16 ``, it will skip the comparison with "utf8" when the first two are compared.
+
+If we add a wild catch to match any encoding
+
+```reasonml
+let encoding = (enc) =>
+  switch (enc) {
+  | `utf8 => 0
+  | `ascii => 1
+  | `utf16 => 2
+  | _ => 3
+  };
+```
+
+It will generate JS as below:
+
+```js
+function encoding(en) {
+  if (en === "utf8") {
+    return 0;
+  } else if (en === "ascii") {
+    return 1;
+  } else if (en === "utf16") {
+    return 2;
+  } else {
+    return 3;
+  }
+}
+```
+
+## Declaring types for string literal types
+
+Note all string literal types can be inferred, this is great for develop experience when you are doing the development. When things get more stable, it would be nice to give string literal types a name as below:
```

Comment:
Note that all string literal types can be inferred. This is very convenient for you when you are doing development.